torchrl/envs/transforms/transforms.py

```diff
@@ -8557,3 +8557,157 @@ def _inv_call(self, tensordict):
         if self.sampling == self.SamplingStrategy.RANDOM:
             action = action + self.jitters * torch.rand_like(self.jitters)
         return tensordict.set(self.in_keys_inv[0], action)
+
+
+class AbsorbingStateTransform(ObservationTransform):
+    """Adds an absorbing state to the observation space.
+
+    A transform that introduces an absorbing state to the environment. This absorbing state is typically used
+    in reinforcement learning to handle terminal states by creating an additional state that signifies
+    the end of an episode but allows for additional steps in the transition to better handle
+    learning algorithms.
+
+    Args:
+        max_episode_length (int): Maximum length of an episode.
+        in_keys (Sequence[NestedKey], optional): Keys to use for input observation. Defaults to ``"observation"``.
+        out_keys (Sequence[NestedKey], optional): Keys to use for output observation. Defaults to ``in_keys``.
+        done_key (Optional[NestedKey]): Key indicating if the episode is done. Defaults to ``"done"``.
+        terminate_key (Optional[NestedKey]): Key indicating if the episode is terminated. Defaults to ``"terminated"``.
+
+    Examples:
+        >>> from torchrl.envs import GymEnv
+        >>> t = AbsorbingStateTransform(max_episode_length=1000)
+        >>> base_env = GymEnv("HalfCheetah-v4")
+        >>> env = TransformedEnv(base_env, t)
+    """
+
+    def __init__(
+        self,
+        max_episode_length: int,
+        in_keys: Sequence[NestedKey] | None = None,
+        out_keys: Sequence[NestedKey] | None = None,
+        done_key: Optional[NestedKey] = "done",
+        terminate_key: Optional[NestedKey] = "terminated",
+    ):
+        if in_keys is None:
+            in_keys = "observation"  # default
+        if out_keys is None:
+            out_keys = copy(in_keys)
+        super().__init__(in_keys=in_keys, out_keys=out_keys)
+        self.max_episode_length = max_episode_length
+        self.done_key = done_key
+        self.terminate_key = terminate_key
+        self._done = None
+        self._curr_timestep = 0
+
+    def forward(self, tensordict: TensorDictBase) -> TensorDictBase:
+        raise RuntimeError(FORWARD_NOT_IMPLEMENTED.format(type(self)))
+
+    def _apply_transform(self, observation: torch.Tensor) -> torch.Tensor:
```

Comment:
there is a version of this that works for all batch sizes. This one will only work with uni of bidimensional batch sizes.